### PR TITLE
[MM-54819] Convert ./components/external_image/external_image.tsx from Class Component to Function Component

### PR DESCRIPTION
**Pull request by gibsonliketheguitar**
#### Summary

- Refactors webapp/channels/src/components/external_image.tsx from class component to functional component.
- Separate isSVGImage function into it's own module -> ```external_image_isSVGImage.ts```
- Refactor and separate isSVGImage test from 
```external_image.test.tsx``` -> ```external_image_isSVGImage.test.tsx```

#### Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/24780
Jira https://mattermost.atlassian.net/browse/MM-54819

#### Release Note

```release-note
* Refactor class to function component
* Separate isSVGImage into its own module
```

### CODE REVIEW

File: webapp/channels/src/components/external_image/external_image.tsx [hunk 1]
```diff
@@ -1,43 +1,30 @@
...
-    render() {
-        let src = getImageSrc(this.props.src, this.props.hasImageProxy);
-
-        if (!this.shouldRenderImage()) {
-            src = '';
-        }
-
-        return this.props.children(src);
+const ExternalImage = (props: Props) => {
+    const shouldRenderImage = () => props.enableSVGs || !isSVGImage(props.imageMetadata, props.src);
```
Comment (larkox) [inline webapp/channels/src/components/external_image/external_image.tsx:22]: This doesn't need to be a function anymore. Also, on line 24, you are not calling it, so it would be giving some trouble. If you make this a simple variable it would be ok.

File: webapp/channels/src/components/external_image/external_image.tsx [hunk 2]
```diff
@@ -1,43 +1,30 @@
...
 
 import type {PostImage} from '@mattermost/types/posts';
 
 import {getImageSrc} from 'utils/post_utils';
 
+import {isSVGImage} from './external_image_isSVGImage';
+
 interface Props {
-    children: (src: string) => React.ReactNode;
+    children: (src: string) => ReactElement | null;
```
Comment (larkox) [inline webapp/channels/src/components/external_image/external_image.tsx:14]: Why this change? `React.ReactNode` should be able to handle null.

File: webapp/channels/src/components/external_image/external_image_isSVGImage.ts
Comment (larkox) [inline webapp/channels/src/components/external_image/external_image_isSVGImage.ts:1]: Nice!

File: webapp/channels/src/components/external_image/external_image.test.tsx [hunk 3]
```diff
@@ -23,7 +23,7 @@ describe('ExternalImage', () => {
     };
 
     test('should render an image', () => {
-        const wrapper = shallow<ExternalImage>(<ExternalImage {...baseProps}/>);
+        const wrapper = shallow<typeof ExternalImage>(<ExternalImage {...baseProps}/>);
```
Comment (larkox) [inline webapp/channels/src/components/external_image/external_image.test.tsx:26]: I think you won't need the typing (and I think it is not 100% correct, but not sure).
```suggestion
        const wrapper = shallow(<ExternalImage {...baseProps}/>);
```

**Review by larkox — CHANGES_REQUESTED** [review]
Great work! A few comments.

File: webapp/channels/src/components/external_image/external_image.tsx (hunk 2 above)
Comment (gibsonliketheguitar) [reply webapp/channels/src/components/external_image/external_image.tsx:13]: Initially, I was getting type errors because I was using memo. When I replaced it with ReactElement, it fixed the ts error.

However when I ran ```make test``` Snapshot test in other component had External Image was failing. 

I have since reverted it back. to React.ReactNode

File: webapp/channels/src/components/external_image/external_image.test.tsx (hunk 3 above)
Comment (gibsonliketheguitar) [reply webapp/channels/src/components/external_image/external_image.test.tsx:26]: I am getting this error when I remove typeof

```
ExternalImage' refers to a value, but is being used as a type here. Did you mean 'typeof ExternalImage'?ts(2749)

```

Comment (larkox) [reply webapp/channels/src/components/external_image/external_image.test.tsx:26]: Yeah, my proposal is to remove everything inside the `<>`, not only the typeof.

File: webapp/channels/src/components/external_image/external_image.tsx (hunk 2 above)
Comment (larkox) [reply webapp/channels/src/components/external_image/external_image.tsx]: Oh,  I see. Then it should be as simple as to wrap the returned children in a react fragment. Something like...
```
return (<>
    {props.children(src);}
</>);
```

File: webapp/channels/src/components/external_image/external_image.tsx [hunk 4]
```diff
@@ -23,7 +23,7 @@ const ExternalImage = (props: Props) => {
     if (!shouldRenderImage) {
         src = '';
     }
-    return props.children(src);
+    return (<>{props.children(src)}</>);
 };
 
-export default ExternalImage;
+export default memo(ExternalImage);
```
Comment (gibsonliketheguitar) [inline webapp/channels/src/components/external_image/external_image.tsx:29]: I added back memo. Is there a script to update snapshots or do i need to manually edit it?

Screenshot 2023-10-18 at 02 23 03

I checked the mattermost channel. The only things that came up was https://community.mattermost.com/core/pl/s6bx66zgjinhfdndnwrwa18w7r

File: webapp/channels/src/components/external_image/external_image.tsx (hunk 2 above)
Comment (gibsonliketheguitar) [reply webapp/channels/src/components/external_image/external_image.tsx]: nice. changes made

File: webapp/channels/src/components/external_image/external_image.test.tsx (hunk 3 above)
Comment (gibsonliketheguitar) [reply webapp/channels/src/components/external_image/external_image.test.tsx]: My b. Brain was not working. Changes made

File: webapp/channels/src/components/external_image/external_image.tsx (hunk 4 above)
Comment (larkox) [reply webapp/channels/src/components/external_image/external_image.tsx:29]: You can run `npm run test -- -u -- path/to/test.test.ts`

Comment (gibsonliketheguitar) [reply webapp/channels/src/components/external_image/external_image.tsx:29]: ty ty

**Review by larkox — APPROVED** [review]
Looks great. Thanks!

File: webapp/channels/src/components/external_image/external_image_isSVGImage.ts
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/external_image/external_image_isSVGImage.ts:1]: nit filename should be `webapp/channels/src/components/external_image/is_svg_image.ts`
We follow this naming case

File: webapp/channels/src/components/external_image/external_image_isSVGImage.test.tsx
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/external_image/external_image_isSVGImage.test.tsx:1]: similar naming case suggestion

File: webapp/channels/src/components/post_view/message_attachments/message_attachment/__snapshots__/message_attachment.test.tsx.snap [hunk 5]
```diff
@@ -29,12 +29,12 @@ exports[`components/post_view/MessageAttachment should call actions.doPostAction
         key="attachment__author-name"
         location="message_attachment"
       >
-        <Connect(ExternalImage)
+        <Connect(Component)
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/post_view/message_attachments/message_attachment/__snapshots__/message_attachment.test.tsx.snap:32]: I am not worried about this but i think this happened because of memo @larkox do you think adding display name is a good idea

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/post_view/message_attachments/message_attachment/__snapshots__/message_attachment.test.tsx.snap (hunk 5 above)
Comment (larkox) [reply webapp/channels/src/components/post_view/message_attachments/message_attachment/__snapshots__/message_attachment.test.tsx.snap:32]: I want to shy away from shallow snapshot testing at some point. That is why I haven't given it more thought.

Also, I think this is an artifact of shallow rendering. Keep in mind that you are not passing the component to the connector, but the memo. So the shallow rendering has not gone deep enough there. In order for this to show the correct name I think we should do something like:
```
const memoizedComponent = React.memo(Component);
memoizedComponent.displayName = 'SomeName'
export default memoizedComponent;
```
Which is a lot of boilerplate to just make the shallow tests happy.

Comment (M-ZubairAhmed) [reply webapp/channels/src/components/post_view/message_attachments/message_attachment/__snapshots__/message_attachment.test.tsx.snap:32]: Yes i agree on your point

**Review by M-ZubairAhmed — APPROVED** [review]
Nice work @gibsonliketheguitar

**Comment by gibsonliketheguitar** [comment]
@M-ZubairAhmed when should I rebase and squash comments?

**Comment by M-ZubairAhmed** [comment]
> @M-ZubairAhmed when should I rebase and squash comments?

I don't see any conflicts but i can do it for my with the command

**Comment by gibsonliketheguitar** [comment]
> > @M-ZubairAhmed when should I rebase and squash comments?
> 
> I don't see any conflicts but i can do it for my with the command

Sounds good

**Review by lindy65 — APPROVED** [review]
Thanks @gibsonliketheguitar - looks good to me 👍 

cc @M-ZubairAhmed
